### PR TITLE
Upgrade jms serializer to newest version

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes -
| Related issues/PRs | Requires: https://github.com/massiveart/MassiveSearchBundle/pull/132
| License | MIT
| Documentation PR | -

#### What's in this PR?

Upgrade jms serializer to newest version

#### Why?

To make possible that the sulu devs can use the newest dependencies.

#### To Do

- [x] Add breaking changes to UPGRADE.md

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -30,17 +30,17 @@
...
         "guzzlehttp/psr7": "^1.5",
         "imagine/imagine": "^0.6.1 || ^0.7.0 || ^1.0",
-        "jms/serializer": "^1.13",
-        "jms/serializer-bundle": "^2.3",
+        "jms/serializer": "^3.0",
+        "jms/serializer-bundle": "^3.0",
         "layershifter/tld-extract": "^1.1 || ^2.0",
         "massive/build-bundle": "^0.3 || ^0.4",
-        "massive/search-bundle": "^0.16 || ^0.17 || ^1.0",
+        "massive/search-bundle": "dev-feature/dependency-update",
```
Comment (alexander-schranz) [inline composer.json]: Need to be merged and tagged as 2.0 before this PR can be merged.
https://github.com/massiveart/MassiveSearchBundle/pull/132

File: src/Sulu/Bundle/AdminBundle/Serializer/Handler/RepresentationHandler.php [hunk 2]
```diff
@@ -11,54 +11,47 @@
...
 use Sulu\Component\Rest\ListBuilder\RepresentationInterface;
-use Sulu\Component\SmartContent\Rest\ItemCollectionRepresentation;
 
 /**
  * @internal
  *
  * This handler workaround some problems with serialize Representation in specific groups
  */
-class RepresentationHandler implements SubscribingHandlerInterface
+class RepresentationHandler implements EventSubscriberInterface
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Serializer/Handler/RepresentationHandler.php]: Shouldn't that class be renamed to `Subscriber` then?

File: src/Sulu/Bundle/CustomUrlBundle/EventListener/CustomUrlSerializeEventSubscriber.php [hunk 3]
```diff
@@ -61,23 +63,42 @@ public static function getSubscribedEvents()
...
+
+        /** @var SerializationVisitorInterface $visitor */
         $visitor = $event->getVisitor();
 
         if (!$customUrl instanceof CustomUrlDocument) {
             return;
         }
 
         if (null !== $customUrl->getTargetDocument()) {
-            $visitor->addData('targetTitle', $customUrl->getTargetDocument()->getTitle());
```
Comment (danrot) [inline src/Sulu/Bundle/CustomUrlBundle/EventListener/CustomUrlSerializeEventSubscriber.php]: Does `addData` not exist anymore? The new way of doing this is very verbose :thinking:

File: src/Sulu/Bundle/MediaBundle/Search/Subscriber/MediaSearchSubscriber.php [hunk 4]
```diff
@@ -93,10 +93,9 @@ public static function getSubscribedEvents()
     public function handlePreIndex(PreIndexEvent $event)
     {
         $metadata = $event->getMetadata();
-        $reflection = $metadata->getClassMetadata()->reflection;
 
         if (
-            false === $reflection->isSubclassOf(FileVersionMeta::class)
+            !$event->getSubject() instanceof FileVersionMeta
             && FileVersionMeta::class !== $metadata->getName()) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Search/Subscriber/MediaSearchSubscriber.php]: Can you move the closing bracket from the if into a separate line? This way it reads a little bit strange.

File: src/Sulu/Bundle/MediaBundle/Search/Subscriber/StructureMediaSearchSubscriber.php [hunk 5]
```diff
@@ -81,7 +81,9 @@ public function handlePreIndex(PreIndexEvent $e)
         $subject = $e->getSubject();
         $evaluator = $e->getFieldEvaluator();
 
-        if (false === $metadata->getClassMetadata()->reflection->isSubclassOf(StructureBehavior::class)) {
+        if (
+            !$e->getSubject() instanceof StructureBehavior
+            && StructureBehavior::class !== $metadata->getName()) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Search/Subscriber/StructureMediaSearchSubscriber.php]: Can you move the closing bracket from the if into a separate line? This way it reads a little bit strange.

File: src/Sulu/Component/Content/Compat/Serializer/PageBridgeHandler.php [hunk 6]
```diff
@@ -23,7 +27,7 @@
 /**
  * Handle serialization and deserialization of the PageBridge.
  */
-class PageBridgeHandler implements SubscribingHandlerInterface
+class PageBridgeHandler implements SubscribingHandlerInterface, EventSubscriberInterface
```
Comment (danrot) [inline src/Sulu/Component/Content/Compat/Serializer/PageBridgeHandler.php]: Does mixing both interfaces feel right to you? Confused me at first :thinking:

File: src/Sulu/Component/Rest/ListBuilder/CollectionRepresentation.php [hunk 7]
```diff
@@ -49,6 +53,9 @@ public function getRel(): string
         return $this->rel;
     }
 
+    /**
+     * @return array
```
Comment (danrot) [inline src/Sulu/Component/Rest/ListBuilder/CollectionRepresentation.php]: Why? The type annotation is already there in the function signature.

File: src/Sulu/Bundle/CustomUrlBundle/EventListener/CustomUrlSerializeEventSubscriber.php (hunk 3 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/CustomUrlBundle/EventListener/CustomUrlSerializeEventSubscriber.php]: Sadly yes. I also don't like it:

See https://github.com/schmittjoh/serializer/blob/master/UPGRADING.md

> Removed deprecated method JsonSerializationVisitor::addData, use ::visitProperty(new StaticPropertyMetadata('', 'name', 'value'), 'value') instead

We could add `trait` or something to wrap this.

File: src/Sulu/Component/Content/Compat/Serializer/PageBridgeHandler.php (hunk 6 above)
Comment (alexander-schranz) [reply src/Sulu/Component/Content/Compat/Serializer/PageBridgeHandler.php]: Done created an own PageBridgeSubscriber

File: src/Sulu/Bundle/AdminBundle/Serializer/Handler/RepresentationHandler.php (hunk 2 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/AdminBundle/Serializer/Handler/RepresentationHandler.php]: Done.

**Comment by danrot** [comment]
That also requires some changes in `sulu-skeleton`, doesn't it? I get some configuration errors.

**Pull request merged**